### PR TITLE
indexOf overloads for matching a series of bytes.

### PR DESCRIPTION
**Pull request by JakeWharton**

### CODE REVIEW

File: okio/src/main/java/okio/Buffer.java [hunk 1]
```diff
@@ -1253,6 +1253,28 @@ Segment writableSegment(int minimumCapacity) {
...
   }
 
+  @Override public long indexOf(ByteString bytes) throws IOException {
+    return indexOf(bytes, 0);
+  }
+
+  @Override public long indexOf(ByteString bytes, long fromIndex) throws IOException {
+    int bytesSize = bytes.size();
+    loop:
+    while ((fromIndex = indexOf(bytes.getByte(0), fromIndex)) != -1) {
```
Comment (JakeWharton) [inline okio/src/main/java/okio/Buffer.java]: If you think there's too much going on in this line I can switch to a `while (true)` loop and move it to an `if` check. That was in my original implementation.

Comment (chrisjenx) [inline okio/src/main/java/okio/Buffer.java]: IMO, I think it is easier to read the previous way.

File: okio/src/main/java/okio/Buffer.java [hunk 2]
```diff
@@ -1253,6 +1253,28 @@ Segment writableSegment(int minimumCapacity) {
     return -1L;
   }
 
+  @Override public long indexOf(ByteString bytes) throws IOException {
+    return indexOf(bytes, 0);
+  }
+
+  @Override public long indexOf(ByteString bytes, long fromIndex) throws IOException {
```
Comment (swankjesse) [inline okio/src/main/java/okio/Buffer.java]: crashes if `bytes.size() == 0`

File: okio/src/main/java/okio/Buffer.java [hunk 3]
```diff
@@ -1253,6 +1253,28 @@ Segment writableSegment(int minimumCapacity) {
...
+    int bytesSize = bytes.size();
+    loop:
+    while ((fromIndex = indexOf(bytes.getByte(0), fromIndex)) != -1) {
+      if (!request(fromIndex + bytesSize)) {
+        return -1; // Not enough bytes to match.
+      }
+      for (int i = 1; i < bytesSize; i++) {
```
Comment (swankjesse) [inline okio/src/main/java/okio/Buffer.java]: I'm guessing we want a helper method like this:
```
public boolean rangeMatches(int offset, ByteString other, int otherOffset, int byteCount);
```
That would be this loop!

Comment (JakeWharton) [inline okio/src/main/java/okio/Buffer.java]: Yeah. I thought the same in doing this.

File: okio/src/main/java/okio/Buffer.java [hunk 4]
```diff
@@ -1253,6 +1253,28 @@ Segment writableSegment(int minimumCapacity) {
...
 
+  @Override public long indexOf(ByteString bytes) throws IOException {
+    return indexOf(bytes, 0);
+  }
+
+  @Override public long indexOf(ByteString bytes, long fromIndex) throws IOException {
+    int bytesSize = bytes.size();
+    loop:
+    while ((fromIndex = indexOf(bytes.getByte(0), fromIndex)) != -1) {
+      if (!request(fromIndex + bytesSize)) {
```
Comment (swankjesse) [inline okio/src/main/java/okio/Buffer.java]: If you have a helper regionMatches method, that should take care of this check

File: okio/src/main/java/okio/BufferedSource.java [hunk 5]
```diff
@@ -215,6 +215,21 @@
   long indexOf(byte b, long fromIndex) throws IOException;
 
   /**
+   * Returns the index of the first match for {@code bytes} in the buffer. This expands the buffer
```
Comment (swankjesse) [inline okio/src/main/java/okio/BufferedSource.java]: Awesome.

File: okio/src/test/java/okio/BufferedSourceTest.java [hunk 6]
```diff
@@ -463,6 +463,26 @@
     assertEquals(15, source.indexOf((byte) 'b', 15));
   }
 
+  @Test public void indexOfByteString() throws IOException {
+    sink.writeUtf8("abacad");
+    assertEquals(2, source.indexOf(ByteString.encodeUtf8("ac")));
+    source.readUtf8(); // Clear stream.
+
+    // Make sure we backtrack and resume searching after partial match.
```
Comment (swankjesse) [inline okio/src/test/java/okio/BufferedSourceTest.java]: Nice test.

**Comment by swankjesse** [comment]
Nice. Ping me when I should take another look!

File: okio/src/main/java/okio/Buffer.java (hunk 3 above)
Comment (JakeWharton) [inline okio/src/main/java/okio/Buffer.java]: Turns out we have this on `ByteString` for matching others called `rangeEquals`.

**Comment by JakeWharton** [comment]
PTAL. I'm holding off on adding a public `rangeEquals` because it's complicated to do efficiently on the underlying `byte[]`s. Plus I'd like the operator additions to be atomic in the git history.

**Comment by swankjesse** [comment]
Nice!

**Comment by chrisjenx** [comment]
:+1:

**Pull request merged**